src/renderers/WebGLRenderer.js

```diff
@@ -1011,6 +1011,53 @@ function WebGLRenderer( parameters ) {
 
 	}
 
+	this.prepareMaterialsAndPrograms = function ( scene, camera, lights, shadows, objects ) {
```

Comment:
The only difference between this and compileAndUploadMaterials is that the ladder traverses the scenes to find the necessary data, where-as this method takes them as arguments. This means that we can pass objects into this method that are not yet in the scene.

Reply:
I think I should replace `compileAndUploadMaterials` with `prepareMaterialsAndPrograms` (or some other, shorter name)